examples/ray/basic_ray_elastic.py

```diff
@@ -0,0 +1,346 @@
+    # issues with Infiniband implementations that are not fork-safe
+    kwargs = {'num_workers': 0, 'pin_memory': True} if args.cuda else {}
+    from filelock import FileLock
+    with FileLock(os.path.expanduser("~/.datalock")):
+        train_dataset = \
+            datasets.MNIST('./data', train=True, download=True,
+                           transform=transforms.Compose([
+                               transforms.ToTensor(),
+                               transforms.Normalize((0.1307,), (0.3081,))
+                           ]))
+    train_sampler = ElasticSampler(train_dataset)
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=4,
+        sampler=train_sampler, **kwargs)
+
+    return train_loader, train_sampler
+
+
+def load_data_cifar():
+    # Horovod: limit # of CPU threads to be used per worker.
+    torch.set_num_threads(4)
+
+    # When supported, use 'forkserver' to spawn dataloader workers instead of 'fork' to prevent
+    # issues with Infiniband implementations that are not fork-safe
+    kwargs = {'num_workers': 0, 'pin_memory': True} if args.cuda else {}
+    # if (kwargs.get('num_workers', 0) > 0 and hasattr(mp, '_supports_context') and
+    #         mp._supports_context and 'forkserver' in mp.get_all_start_methods()):
+    # kwargs['multiprocessing_context'] = 'spawn'
+    transform_train = transforms.Compose([
+        transforms.RandomCrop(32, padding=4),
+        transforms.RandomHorizontalFlip(),
+        transforms.ToTensor(),
+        transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
+    ])
+
+    from filelock import FileLock
+    with FileLock(os.path.expanduser("~/.datalock")):
+        train_dataset = datasets.CIFAR10(
+            root=args.data_dir, train=True, download=True, transform=transform_train)
+    train_sampler = ElasticSampler(train_dataset)
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=4,
+        sampler=train_sampler, **kwargs)
+    return train_loader, train_sampler
+
+
+class tqdm_callback:
+    def __init__(self):
+        self._progress_bar = None
+        self._current_epoch = None
+        self._mode = None
+
+    def __call__(self, info):
+        tqdm_mode = info["tqdm_mode"]
+        assert tqdm_mode in {"val", "train"}
+        reset = False
+        if self._mode != tqdm_mode or self._current_epoch != info["epoch"]:
+            reset = True
+            self._mode = tqdm_mode
+            self._current_epoch = info["epoch"]
+
+        if reset:
+            if self._progress_bar is not None:
+                self._progress_bar.close()
+
+            self._progress_bar = tqdm(
+                total=info["total"],
+                desc=f'[mode={tqdm_mode}] Epoch     #{self._current_epoch + 1}')
+
+        scoped = {k: v for k, v in info.items() if k.startswith(tqdm_mode)}
+        self._progress_bar.set_postfix(scoped)
+        self._progress_bar.update(1)
```

Comment:
Should we also create a callback with a `SummaryWriter` that logs the `info` to the log directory?

Reply:
done!